File: src/sc4pimx/FSHReader.py
"""FSH (Maxis Texture) file reader for SC4PIM.

This module handles reading and parsing of FSH texture files used in SimCity 4.
Based on the C++ implementation in FSHReader.cpp.
"""

import struct
from dataclasses import dataclass
from typing import List, NamedTuple, Tuple

import numpy as np

from . import QFS


class Bitmap(NamedTuple):
    """Represents a single bitmap/mipmap level in an FSH entry."""

    code: int
    width: int
    height: int
    mip_level: int
    data: bytes


@dataclass
class Entry:
    """Represents a single entry (texture) in an FSH file."""

    name: str
    format_code: int
    width: int
    height: int
    mip_count: int
    bitmaps: List[Bitmap]
    label: str = ""


@dataclass
class FileHeader:
    """FSH file header."""

    magic: int
    size: int
    num_entries: int
    dir_id: int

    def is_valid(self) -> bool:
        """Check if header is valid."""
        return self.magic in {
            0x49504853,  # 'SHPI'
            0x34363247,  # 'G264'
            0x36363247,  # 'G266'
            0x34353347,  # 'G354'
        }


@dataclass
class Record:
    """Represents a complete FSH file."""

    header: FileHeader
    entries: List[Entry]


class FSHReader:
    """Reader for FSH (Maxis Texture) files."""

    # Format codes
    CODE_32BIT = 0x7D
    CODE_24BIT = 0x7F
    CODE_4444 = 0x6D
    CODE_0565 = 0x78
    CODE_1555 = 0x7E
    CODE_DXT1 = 0x60
    CODE_DXT3 = 0x61
    CODE_DXT5 = 0x62

    @staticmethod
    def parse(buffer: bytes) -> Record:
        """Parse an FSH file.

        Args:
            buffer: Raw FSH file data (may be QFS compressed)

        Returns:
            Record object containing parsed FSH data

        Raises:
            ValueError: If parsing fails
        """
        if len(buffer) < 16:
            raise ValueError("Buffer too small for FSH header")

        # Decompress if needed
        file_span = buffer
        if QFS.is_qfs_compressed(buffer):
            try:
                file_span = QFS.decode(buffer)
                if file_span is None:
                    raise ValueError("Invalid QFS stream")
            except ValueError as e:
                raise ValueError(f"Failed to decompress FSH payload: {e}") from e

        reader = _SpanReader(file_span)

        # Read header
        magic = reader.read_le_u32()
        size = reader.read_le_u32()
        num_entries = reader.read_le_u32()
        dir_id = reader.read_le_u32()

        header = FileHeader(magic=magic, size=size, num_entries=num_entries, dir_id=dir_id)

        if not header.is_valid():
            raise ValueError("Invalid FSH header")

        # Read directory entries
        directory: List[Tuple[str, int]] = []
        for i in range(num_entries):
            name_bytes = reader.read_bytes(4)
            name = _make_name(name_bytes)
            offset = reader.read_le_u32()
            directory.append((name, offset))

        # Parse entries
        entries: List[Entry] = []
        for i in range(num_entries):
            name, offset = directory[i]
            next_offset = (
                directory[i + 1][1] if i + 1 < len(directory) else len(file_span)
            )

            if offset >= len(file_span) or offset >= next_offset:
                raise ValueError("Invalid FSH directory offsets")

            entry_data = file_span[offset:next_offset]
            entry_reader = _SpanReader(entry_data)

            entry = Entry(
                name=name, format_code=0, width=0, height=0, mip_count=0, bitmaps=[]
            )

            record_byte = entry_reader.read_u8()
            block_size = _read_u24le(entry_reader)

            width = entry_reader.read_le_u16()
            height = entry_reader.read_le_u16()
            _x_center = entry_reader.read_le_u16()
            _y_center = entry_reader.read_le_u16()
            _x_offset = entry_reader.read_le_u16()
            y_offset = entry_reader.read_le_u16()

            entry.format_code = record_byte & 0x7F
            entry.width = width
            entry.height = height
            entry.mip_count = (y_offset >> 12) & 0x0F

            # Read bitmaps/mipmaps
            for mip in range(entry.mip_count + 1):
                mip_width = max(1, width >> mip)
                mip_height = max(1, height >> mip)

                # Skip if DXT and dimensions not multiple of 4
                if (
                    entry.format_code in (FSHReader.CODE_DXT1, FSHReader.CODE_DXT3)
                    and (mip_width % 4 != 0 or mip_height % 4 != 0)
                ):
                    break

                bitmap = Bitmap(
                    code=entry.format_code,
                    width=mip_width,
                    height=mip_height,
                    mip_level=mip,
                    data=b"",
                )

                data_size = FSHReader._expected_data_size(bitmap.code, bitmap.width, bitmap.height)
                bitmap_data = entry_reader.read_bytes(data_size)
                bitmap = Bitmap(
                    code=bitmap.code,
                    width=bitmap.width,
                    height=bitmap.height,
                    mip_level=bitmap.mip_level,
                    data=bitmap_data,
                )
                entry.bitmaps.append(bitmap)

            # Read label if present
            if block_size != 0:
                attachment_offset = offset + block_size
                if attachment_offset + 4 < next_offset:
                    attachment_data = file_span[attachment_offset : next_offset]
                    if len(attachment_data) >= 5 and attachment_data[0] == 0x70:
                        label_start = 4
                        label_end = attachment_data.find(b"\x00", label_start)
                        if label_end == -1:
                            label_end = len(attachment_data)
                        entry.label = attachment_data[label_start:label_end].decode(
                            "utf-8", errors="ignore"
                        )

            entries.append(entry)

        return Record(header=header, entries=entries)

    @staticmethod
    def convert_to_rgba8(bitmap: Bitmap) -> bytes:
        """Convert bitmap to RGBA8 format.

        Args:
            bitmap: Bitmap to convert

        Returns:
            RGBA8 pixel data (4 bytes per pixel)

        Raises:
            ValueError: If conversion fails
        """
        if bitmap.width == 0 or bitmap.height == 0:
            raise ValueError("Invalid bitmap dimensions")

        expected_size = FSHReader._expected_data_size(
            bitmap.code, bitmap.width, bitmap.height
        )
        if expected_size == 0:
            raise ValueError(f"Unsupported format code: {bitmap.code}")
        if len(bitmap.data) != expected_size:
            raise ValueError(
                f"Bitmap data size mismatch (expected {expected_size}, got {len(bitmap.data)})"
            )

        code = bitmap.code
        if code in (FSHReader.CODE_DXT1, FSHReader.CODE_DXT3, FSHReader.CODE_DXT5):
            if bitmap.width % 4 != 0 or bitmap.height % 4 != 0:
                raise ValueError("DXT dimensions must be multiple of 4")
            return FSHReader._convert_dxt(bitmap)
        if code == FSHReader.CODE_32BIT:
            return FSHReader._convert_32bit(bitmap.data)
        if code == FSHReader.CODE_24BIT:
            return FSHReader._convert_24bit(bitmap.data)
        if code == FSHReader.CODE_4444:
            return FSHReader._convert_4444(bitmap.data)
        if code == FSHReader.CODE_0565:
            return FSHReader._convert_0565(bitmap.data)
        if code == FSHReader.CODE_1555:
            return FSHReader._convert_1555(bitmap.data)
        raise ValueError(f"Unsupported format code: {code}")

    # The conversions below are vectorised with numpy: the previous
    # pure-Python per-pixel loops cost ~10-20 ms per 256x256 texture, which
    # dominated preview/ImageDB generation. numpy brings that to well under
    # 1 ms while producing byte-identical output.

    @staticmethod
    def _convert_32bit(src: bytes) -> bytes:
        """Convert 32-bit BGRA to RGBA8."""
        arr = np.frombuffer(src, dtype=np.uint8).reshape(-1, 4)
        return np.ascontiguousarray(arr[:, [2, 1, 0, 3]]).tobytes()

    @staticmethod
    def _convert_24bit(src: bytes) -> bytes:
        """Convert 24-bit BGR to RGBA8."""
        arr = np.frombuffer(src, dtype=np.uint8).reshape(-1, 3)
        out = np.empty((arr.shape[0], 4), dtype=np.uint8)
        out[:, 0] = arr[:, 2]
        out[:, 1] = arr[:, 1]
        out[:, 2] = arr[:, 0]
        out[:, 3] = 255
        return out.tobytes()

    @staticmethod
    def _convert_4444(src: bytes) -> bytes:
        """Convert ARGB4444 to RGBA8."""
        v = np.frombuffer(src, dtype="<u2").astype(np.uint32)
        out = np.empty((v.shape[0], 4), dtype=np.uint8)
        a = (v >> 12) & 0xF
        r = (v >> 8) & 0xF
        g = (v >> 4) & 0xF
        b = v & 0xF
        out[:, 0] = (r << 4) | r
        out[:, 1] = (g << 4) | g
        out[:, 2] = (b << 4) | b
        out[:, 3] = (a << 4) | a
        return out.tobytes()

    @staticmethod
    def _convert_0565(src: bytes) -> bytes:
        """Convert RGB565 to RGBA8."""
        v = np.frombuffer(src, dtype="<u2").astype(np.uint32)
        out = np.empty((v.shape[0], 4), dtype=np.uint8)
        r = (v >> 11) & 0x1F
        g = (v >> 5) & 0x3F
        b = v & 0x1F
        out[:, 0] = (r << 3) | (r >> 2)
        out[:, 1] = (g << 2) | (g >> 4)
        out[:, 2] = (b << 3) | (b >> 2)
        out[:, 3] = 255
        return out.tobytes()

    @staticmethod
    def _convert_1555(src: bytes) -> bytes:
        """Convert ARGB1555 to RGBA8."""
        v = np.frombuffer(src, dtype="<u2").astype(np.uint32)
        out = np.empty((v.shape[0], 4), dtype=np.uint8)
        a = (v >> 15) & 0x1
        r = (v >> 10) & 0x1F
        g = (v >> 5) & 0x1F
        b = v & 0x1F
        out[:, 0] = (r << 3) | (r >> 2)
        out[:, 1] = (g << 3) | (g >> 2)
        out[:, 2] = (b << 3) | (b >> 2)
        out[:, 3] = np.where(a == 1, 255, 0)
        return out.tobytes()

    @staticmethod
    def _convert_dxt(bitmap: Bitmap) -> bytes:
        """Convert DXT1/DXT3/DXT5 compressed data to RGBA8 (vectorised)."""
        code = bitmap.code
        expected_size = FSHReader._expected_data_size(code, bitmap.width, bitmap.height)
        if len(bitmap.data) != expected_size:
            raise ValueError(
                f"DXT data size mismatch (expected {expected_size}, got {len(bitmap.data)})"
            )

        w, h = bitmap.width, bitmap.height
        bw, bh = w // 4, h // 4
        nblocks = bw * bh
        block_size = 8 if code == FSHReader.CODE_DXT1 else 16
        blocks = np.frombuffer(bitmap.data, dtype=np.uint8).reshape(nblocks, block_size)
        # For DXT3/DXT5 the colour block follows the 8-byte alpha block.
        color_block = blocks if code == FSHReader.CODE_DXT1 else blocks[:, 8:]
        idx = np.arange(nblocks)

        c0 = color_block[:, 0].astype(np.uint16) | (color_block[:, 1].astype(np.uint16) << 8)
        c1 = color_block[:, 2].astype(np.uint16) | (color_block[:, 3].astype(np.uint16) << 8)
        codes = (
            color_block[:, 4].astype(np.uint32)
            | (color_block[:, 5].astype(np.uint32) << 8)
            | (color_block[:, 6].astype(np.uint32) << 16)
            | (color_block[:, 7].astype(np.uint32) << 24)
        )

        def rgb565(c: np.ndarray) -> np.ndarray:
            c = c.astype(np.int32)
            r = (c >> 11) & 0x1F
            g = (c >> 5) & 0x3F
            b = c & 0x1F
            return np.stack(
                [(r << 3) | (r >> 2), (g << 2) | (g >> 4), (b << 3) | (b >> 2)], axis=1
            )

        rgb0 = rgb565(c0)  # (nblocks, 3)
        rgb1 = rgb565(c1)
        # palette: (nblocks, 4 colours, RGBA)
        palette = np.zeros((nblocks, 4, 4), dtype=np.int32)
        palette[:, 0, :3] = rgb0
        palette[:, 1, :3] = rgb1
        palette[:, 0:2, 3] = 255
        four = (code != FSHReader.CODE_DXT1) | (c0.astype(np.int32) > c1.astype(np.int32))
        fm = four[:, None]
        palette[:, 2, :3] = np.where(fm, (2 * rgb0 + rgb1) // 3, (rgb0 + rgb1) // 2)
        palette[:, 2, 3] = 255
        palette[:, 3, :3] = np.where(fm, (rgb0 + 2 * rgb1) // 3, 0)
        palette[:, 3, 3] = np.where(four, 255, 0)

        rgba = np.empty((nblocks, 16, 4), dtype=np.int32)
        for i in range(16):
            sel = (codes >> (2 * i)) & 0x03
            rgba[:, i, :] = palette[idx, sel]

        if code == FSHReader.CODE_DXT3:
            ab = blocks[:, :8].astype(np.uint16)
            nib = np.empty((nblocks, 16), dtype=np.uint16)
            nib[:, 0::2] = ab & 0x0F
            nib[:, 1::2] = (ab >> 4) & 0x0F
            rgba[:, :, 3] = nib * 17
        elif code == FSHReader.CODE_DXT5:
            ablk = blocks[:, :8]
            a0 = ablk[:, 0].astype(np.int32)
            a1 = ablk[:, 1].astype(np.int32)
            pal = np.zeros((nblocks, 8), dtype=np.int32)
            pal[:, 0] = a0
            pal[:, 1] = a1
            gt = a0 > a1
            gt_vals = [
                (6 * a0 + a1) // 7, (5 * a0 + 2 * a1) // 7, (4 * a0 + 3 * a1) // 7,
                (3 * a0 + 4 * a1) // 7, (2 * a0 + 5 * a1) // 7, (a0 + 6 * a1) // 7,
            ]
            le_vals = [
                (4 * a0 + a1) // 5, (3 * a0 + 2 * a1) // 5, (2 * a0 + 3 * a1) // 5,
                (a0 + 4 * a1) // 5, np.zeros_like(a0), np.full_like(a0, 255),
            ]
            for k in range(6):
                pal[:, 2 + k] = np.where(gt, gt_vals[k], le_vals[k])
            bits = np.zeros(nblocks, dtype=np.uint64)
            for k in range(6):
                bits |= ablk[:, 2 + k].astype(np.uint64) << np.uint64(8 * k)
            for i in range(16):
                sel = ((bits >> np.uint64(3 * i)) & np.uint64(0x07)).astype(np.intp)
                rgba[:, i, 3] = pal[idx, sel]

        # (nblocks,16,4) -> (bh,bw,4,4,4) -> (h,w,4)
        img = rgba.reshape(bh, bw, 4, 4, 4).transpose(0, 2, 1, 3, 4).reshape(h, w, 4)
        return img.astype(np.uint8).tobytes()

    @staticmethod
    def _expected_data_size(format_code: int, width: int, height: int) -> int:
        """Calculate expected data size for a bitmap."""
        if format_code == FSHReader.CODE_32BIT:
            return width * height * 4
        elif format_code == FSHReader.CODE_24BIT:
            return width * height * 3
        elif format_code in (
            FSHReader.CODE_4444,
            FSHReader.CODE_0565,
            FSHReader.CODE_1555,
        ):
            return width * height * 2
        elif format_code == FSHReader.CODE_DXT1:
            return ((width + 3) // 4) * ((height + 3) // 4) * 8
        elif format_code in (FSHReader.CODE_DXT3, FSHReader.CODE_DXT5):
            return ((width + 3) // 4) * ((height + 3) // 4) * 16
        else:
            return 0


def _make_name(name_bytes: bytes) -> str:
    """Convert 4-byte name field to string."""
    name_str = name_bytes.decode("utf-8", errors="ignore")
    null_pos = name_str.find("\x00")
    if null_pos != -1:
        name_str = name_str[:null_pos]
    return name_str


def _read_u24le(reader: "_SpanReader") -> int:
    """Read a 24-bit FSH block size."""
    byte0 = reader.read_u8()
    byte1 = reader.read_u8()
    byte2 = reader.read_u8()
    return (byte0 << 16) | (byte1 << 8) | byte2


class _SpanReader:
    """Helper class for reading binary data."""

    def __init__(self, data: bytes):
        self.data = data
        self.pos = 0

    def read_u8(self) -> int:
        """Read unsigned 8-bit integer."""
        if self.pos >= len(self.data):
            raise ValueError("Read past end of buffer")
        val = self.data[self.pos]
        self.pos += 1
        return val

    def read_le_u16(self) -> int:
        """Read little-endian unsigned 16-bit integer."""
        if self.pos + 2 > len(self.data):
            raise ValueError("Read past end of buffer")
        val = struct.unpack("<H", self.data[self.pos : self.pos + 2])[0]
        self.pos += 2
        return val

    def read_le_u32(self) -> int:
        """Read little-endian unsigned 32-bit integer."""
        if self.pos + 4 > len(self.data):
            raise ValueError("Read past end of buffer")
        val = struct.unpack("<I", self.data[self.pos : self.pos + 4])[0]
        self.pos += 4
        return val

    def read_bytes(self, count: int) -> bytes:
        """Read raw bytes."""
        if self.pos + count > len(self.data):
            raise ValueError("Read past end of buffer")
        val = self.data[self.pos : self.pos + count]
        self.pos += count
        return val

    def peek_bytes(self, count: int) -> bytes:
        """Peek at bytes without advancing position."""
        if self.pos + count > len(self.data):
            raise ValueError("Read past end of buffer")
        return self.data[self.pos : self.pos + count]

    def skip(self, count: int) -> None:
        """Skip bytes."""
        if self.pos + count > len(self.data):
            raise ValueError("Skip past end of buffer")
        self.pos += count

File: .ai/extra-info/FSHReader.cpp
#include "FSHReader.h"

#include <algorithm>
#include <string>
#include <squish/squish.h>

#include "QFSDecompressor.h"
#include "SafeSpanReader.h"

namespace {

ParseExpected<uint32_t> ReadUInt24(DBPF::SafeSpanReader& reader) {
    auto byte0 = reader.ReadLE<uint8_t>();
    if (!byte0) return std::unexpected(byte0.error());
    auto byte1 = reader.ReadLE<uint8_t>();
    if (!byte1) return std::unexpected(byte1.error());
    auto byte2 = reader.ReadLE<uint8_t>();
    if (!byte2) return std::unexpected(byte2.error());

    uint32_t result = (static_cast<uint32_t>(*byte0) << 16) |
                      (static_cast<uint32_t>(*byte1) << 8) |
                      static_cast<uint32_t>(*byte2);
    return result;
}

std::string MakeName(const char name[4]) {
    std::string s(name, name + 4);
    auto nullPos = s.find('\0');
    if (nullPos != std::string::npos) {
        s.resize(nullPos);
    }
    return s;
}

} // namespace

namespace FSH {

ParseExpected<Record> Reader::Parse(std::span<const uint8_t> buffer) {
    if (buffer.size() < sizeof(FileHeader)) {
        return Fail("Buffer too small for FSH header");
    }

    std::vector<uint8_t> decompressed;
    std::span<const uint8_t> fileSpan = buffer;

    if (QFS::Decompressor::IsQFSCompressed(buffer)) {
        auto result = QFS::Decompressor::Decompress(buffer, decompressed);
        if (!result.has_value()) {
            return Fail("Failed to decompress FSH payload: {}", result.error().message);
        }
        fileSpan = std::span<const uint8_t>(decompressed.data(), decompressed.size());
    }

    DBPF::SafeSpanReader reader(fileSpan);

    Record outFile;
    auto magic = reader.ReadLE<uint32_t>();
    if (!magic) return std::unexpected(magic.error());
    outFile.header.magic = *magic;

    auto size = reader.ReadLE<uint32_t>();
    if (!size) return std::unexpected(size.error());
    outFile.header.size = *size;

    auto numEntries = reader.ReadLE<uint32_t>();
    if (!numEntries) return std::unexpected(numEntries.error());
    outFile.header.numEntries = *numEntries;

    auto dirId = reader.ReadLE<uint32_t>();
    if (!dirId) return std::unexpected(dirId.error());
    outFile.header.dirId = *dirId;

    if (!outFile.header.IsValid()) {
        return Fail("Invalid FSH header");
    }

    struct DirEntryParsed {
        std::string name;
        uint32_t offset{};
    };

    std::vector<DirEntryParsed> directory(outFile.header.numEntries);
    for (uint32_t i = 0; i < outFile.header.numEntries; ++i) {
        DirectoryEntry dir{};
        auto readBytes = reader.ReadBytes(dir.name, sizeof(dir.name));
        if (!readBytes) return std::unexpected(readBytes.error());

        auto offset = reader.ReadLE<uint32_t>();
        if (!offset) return std::unexpected(offset.error());

        directory[i].name = MakeName(dir.name);
        directory[i].offset = *offset;
    }

    outFile.entries.clear();
    outFile.entries.reserve(outFile.header.numEntries);

    for (uint32_t i = 0; i < outFile.header.numEntries; ++i) {
        const uint32_t offset = directory[i].offset;
        const uint32_t nextOffset = (i + 1 < directory.size())
                                        ? directory[i + 1].offset
                                        : static_cast<uint32_t>(fileSpan.size());
        if (offset >= fileSpan.size() || offset >= nextOffset) {
            return Fail("Invalid FSH directory offsets");
        }

        // Create a reader for this entry's data
        auto entrySpan = fileSpan.subspan(offset, nextOffset - offset);
        DBPF::SafeSpanReader entryReader(entrySpan);

        Entry entry{};
        entry.name = directory[i].name;

        auto record = entryReader.ReadLE<uint8_t>();
        if (!record) return std::unexpected(record.error());

        auto blockSize = ReadUInt24(entryReader);
        if (!blockSize) return std::unexpected(blockSize.error());

        auto width = entryReader.ReadLE<uint16_t>();
        if (!width) return std::unexpected(width.error());
        auto height = entryReader.ReadLE<uint16_t>();
        if (!height) return std::unexpected(height.error());
        auto xCenter = entryReader.ReadLE<uint16_t>();
        if (!xCenter) return std::unexpected(xCenter.error());
        auto yCenter = entryReader.ReadLE<uint16_t>();
        if (!yCenter) return std::unexpected(yCenter.error());
        auto xOffset = entryReader.ReadLE<uint16_t>();
        if (!xOffset) return std::unexpected(xOffset.error());
        auto yOffset = entryReader.ReadLE<uint16_t>();
        if (!yOffset) return std::unexpected(yOffset.error());

        entry.formatCode = *record & 0x7F;
        entry.width = *width;
        entry.height = *height;
        entry.mipCount = static_cast<uint8_t>((*yOffset >> 12) & 0x0F);

        for (uint8_t mip = 0; mip <= entry.mipCount; ++mip) {
            uint16_t mipWidth = static_cast<uint16_t>(std::max<int>(1, *width >> mip));
            uint16_t mipHeight = static_cast<uint16_t>(std::max<int>(1, *height >> mip));
            if ((entry.formatCode == kCodeDXT1 || entry.formatCode == kCodeDXT3) &&
                (mipWidth % 4 != 0 || mipHeight % 4 != 0)) {
                break;
            }
            Bitmap bitmap;
            bitmap.code = entry.formatCode;
            bitmap.width = mipWidth;
            bitmap.height = mipHeight;
            bitmap.mipLevel = mip;
            const size_t dataSize = bitmap.ExpectedDataSize();

            auto bitmapData = entryReader.PeekBytes(dataSize);
            if (!bitmapData) return std::unexpected(bitmapData.error());

            bitmap.data.assign(bitmapData->begin(), bitmapData->end());
            auto skip = entryReader.Skip(dataSize);
            if (!skip) return std::unexpected(skip.error());

            entry.bitmaps.push_back(std::move(bitmap));
        }

        if (*blockSize != 0) {
            const uint32_t attachmentOffset = offset + *blockSize;
            if (attachmentOffset + 4 < nextOffset) {
                // Use the original fileSpan to access the attachment
                auto attachmentSpan = fileSpan.subspan(attachmentOffset, nextOffset - attachmentOffset);
                if (attachmentSpan.size() >= 5 && attachmentSpan[0] == 0x70) {
                    auto labelStart = reinterpret_cast<const char*>(attachmentSpan.data() + 4);
                    auto labelEnd = reinterpret_cast<const char*>(attachmentSpan.data() + attachmentSpan.size());
                    const char* terminator = std::find(labelStart, labelEnd, '\0');
                    entry.label.assign(labelStart, terminator);
                }
            }
        }

        outFile.entries.push_back(std::move(entry));
    }

    return outFile;
}

bool Reader::ConvertToRGBA8(const Bitmap& bitmap, std::vector<uint8_t>& outRGBA) {
    if (bitmap.width == 0 || bitmap.height == 0) {
        return false;
    }

    const size_t pixelCount = static_cast<size_t>(bitmap.width) * static_cast<size_t>(bitmap.height);
    outRGBA.assign(pixelCount * 4, 0);

    if (bitmap.IsDXT()) {
        if (bitmap.width % 4 != 0 || bitmap.height % 4 != 0) {
            return false;
        }
    }

    switch (bitmap.code) {
        case kCode32Bit: {
            const uint8_t* src = bitmap.data.data();
            uint8_t* dst = outRGBA.data();
            for (size_t i = 0; i < pixelCount; ++i) {
                uint8_t b = *src++;
                uint8_t g = *src++;
                uint8_t r = *src++;
                uint8_t a = *src++;
                *dst++ = r;
                *dst++ = g;
                *dst++ = b;
                *dst++ = a;
            }
            return true;
        }
        case kCode24Bit: {
            const uint8_t* src = bitmap.data.data();
            uint8_t* dst = outRGBA.data();
            for (size_t i = 0; i < pixelCount; ++i) {
                uint8_t b = *src++;
                uint8_t g = *src++;
                uint8_t r = *src++;
                *dst++ = r;
                *dst++ = g;
                *dst++ = b;
                *dst++ = 255;
            }
            return true;
        }
        case kCode4444: {
            const uint8_t* src = bitmap.data.data();
            uint8_t* dst = outRGBA.data();
            for (size_t i = 0; i < pixelCount; ++i) {
                uint16_t color;
                std::memcpy(&color, src, sizeof(uint16_t));
                src += sizeof(uint16_t);
                Reader::ARGB4444ToRGBA8(color, dst);
                dst += 4;
            }
            return true;
        }
        case kCode0565: {
            const uint8_t* src = bitmap.data.data();
            uint8_t* dst = outRGBA.data();
            for (size_t i = 0; i < pixelCount; ++i) {
                uint16_t color;
                std::memcpy(&color, src, sizeof(uint16_t));
                src += sizeof(uint16_t);
                Reader::RGB565ToRGBA8(color, dst);
                dst += 4;
            }
            return true;
        }
        case kCode1555: {
            const uint8_t* src = bitmap.data.data();
            uint8_t* dst = outRGBA.data();
            for (size_t i = 0; i < pixelCount; ++i) {
                uint16_t color;
                std::memcpy(&color, src, sizeof(uint16_t));
                src += sizeof(uint16_t);
                Reader::ARGB1555ToRGBA8(color, dst);
                dst += 4;
            }
            return true;
        }
        case kCodeDXT1:
        case kCodeDXT3:
        case kCodeDXT5: {
            int squishFlags = squish::kDxt1;
            if (bitmap.code == kCodeDXT3) {
                squishFlags = squish::kDxt3;
            } else if (bitmap.code == kCodeDXT5) {
                squishFlags = squish::kDxt5;
            }
            squish::DecompressImage(outRGBA.data(),
                                    static_cast<int>(bitmap.width),
                                    static_cast<int>(bitmap.height),
                                    bitmap.data.data(),
                                    squishFlags);
            return true;
        }
        default:
            return false;
    }
}

void Reader::ARGB4444ToRGBA8(uint16_t color, uint8_t* rgba) {
    uint8_t a = (color >> 12) & 0xF;
    uint8_t r = (color >> 8) & 0xF;
    uint8_t g = (color >> 4) & 0xF;
    uint8_t b = color & 0xF;
    rgba[0] = static_cast<uint8_t>((r << 4) | r);
    rgba[1] = static_cast<uint8_t>((g << 4) | g);
    rgba[2] = static_cast<uint8_t>((b << 4) | b);
    rgba[3] = static_cast<uint8_t>((a << 4) | a);
}

void Reader::RGB565ToRGBA8(uint16_t color, uint8_t* rgba) {
    uint8_t r = (color >> 11) & 0x1F;
    uint8_t g = (color >> 5) & 0x3F;
    uint8_t b = color & 0x1F;
    rgba[0] = static_cast<uint8_t>((r << 3) | (r >> 2));
    rgba[1] = static_cast<uint8_t>((g << 2) | (g >> 4));
    rgba[2] = static_cast<uint8_t>((b << 3) | (b >> 2));
    rgba[3] = 255;
}

void Reader::ARGB1555ToRGBA8(uint16_t color, uint8_t* rgba) {
    uint8_t a = (color >> 15) & 0x1;
    uint8_t r = (color >> 10) & 0x1F;
    uint8_t g = (color >> 5) & 0x1F;
    uint8_t b = color & 0x1F;
    rgba[0] = static_cast<uint8_t>((r << 3) | (r >> 2));
    rgba[1] = static_cast<uint8_t>((g << 3) | (g >> 2));
    rgba[2] = static_cast<uint8_t>((b << 3) | (b >> 2));
    rgba[3] = a ? 255 : 0;
}

} // namespace FSH
